### PR TITLE
AAE-506 add support to multi-instance result collection

### PR DESCRIPTION
**Pull request by erdemedeiros**
Mapped as `loopDataOutputRef` and `outputDataItem` as defined in the BPMN specification
Ref https://github.com/Activiti/Activiti/issues/3041

### CODE REVIEW

File: activiti-core/activiti-engine/src/main/java/org/activiti/engine/impl/variable/JsonType.java [hunk 1]
```diff
@@ -50,40 +49,25 @@ public boolean isCachable() {
...
-    return jsonValue;
+    return loadedValue;
   }
 
-
   public void setValue(Object value, ValueFields valueFields) {
     try {
       valueFields.setTextValue(objectMapper.writeValueAsString(value));
+      if (value != null) {
+          valueFields.setTextValue2(value.getClass().getName());
```
Comment (erdemedeiros) [inline activiti-core/activiti-engine/src/main/java/org/activiti/engine/impl/variable/JsonType.java]: Here, I'm reusing a field that already exists but that's not used by JSON type (`textValue2`) to store the value of the original Java Class. Is it ok or is it better to create a new column in the database base (`javaType`) for that?

Comment (erdemedeiros) [reply activiti-core/activiti-engine/src/main/java/org/activiti/engine/impl/variable/JsonType.java]: I'll create a specific field: it will be cleaner.

Comment (eromano) [reply activiti-core/activiti-engine/src/main/java/org/activiti/engine/impl/variable/JsonType.java]: I agree

Comment (balsarori) [reply activiti-core/activiti-engine/src/main/java/org/activiti/engine/impl/variable/JsonType.java]: FYI, for JPA variables types textValue and textValue2 were used on for the value and the other for the class name. 
https://github.com/Activiti/Activiti/blob/develop/activiti-core/activiti-engine/src/main/java/org/activiti/engine/impl/variable/JPAEntityVariableType.java#L67

Comment (erdemedeiros) [reply activiti-core/activiti-engine/src/main/java/org/activiti/engine/impl/variable/JsonType.java]: Thanks, @balsarori. Yes, I reached that code as well when I was trying to understand what was the purpose of `textValue2`. It's just hard to know who is who, with such generic names. That's why I was thinking about creating an `originalClassName` attribute.

Comment (erdemedeiros) [reply activiti-core/activiti-engine/src/main/java/org/activiti/engine/impl/variable/JsonType.java]: For simplicity, I'll use `textValue2` without adding a new field.

File: activiti-core/activiti-engine/src/main/java/org/activiti/engine/impl/bpmn/behavior/SequentialMultiInstanceBehavior.java [hunk 2]
```diff
@@ -81,13 +81,16 @@ public void leave(DelegateExecution childExecution) {
     setLoopVariable(multiInstanceRootExecution, NUMBER_OF_COMPLETED_INSTANCES, nrOfCompletedInstances);
     setLoopVariable(childExecution, getCollectionElementIndexVariable(), loopCounter);
     logLoopDetails(childExecution, "instance completed", loopCounter, nrOfCompletedInstances, nrOfActiveInstances, nrOfInstances);
+
+    updateResultCollection(childExecution, multiInstanceRootExecution);
     
     Context.getCommandContext().getHistoryManager().recordActivityEnd((ExecutionEntity) childExecution, null);
     callActivityEndListeners(childExecution);
     
     //executeCompensationBoundaryEvents(execution.getCurrentFlowElement(), execution);
```
Comment (eromano) [inline activiti-core/activiti-engine/src/main/java/org/activiti/engine/impl/bpmn/behavior/SequentialMultiInstanceBehavior.java]: is still needed this commented code?

Comment (erdemedeiros) [reply activiti-core/activiti-engine/src/main/java/org/activiti/engine/impl/bpmn/behavior/SequentialMultiInstanceBehavior.java]: Commented code is never needed :). I removed it.

**Pull request merged**